[pytorch][nnc] update external functions for mobile build

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56852 [pytorch][nnc] mobile nnc backend skeleton
* #56851 [pytorch][nnc] protocol classes to persist the context for compiled functions
* **#56850 [pytorch][nnc] update external functions for mobile build**



This is part of the changes to enable NNC AOT compilation for mobile.
The generated kernels need to call these external functions thus change the declarations to use C linkage when building the mobile runtime.

Added nnc_aten_addmm external function.


Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154/)

Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154)

ljk53 committed:
[pytorch][nnc] update external functions for mobile build

This is part of the changes to enable NNC AOT compilation for mobile.
The generated kernels need to call these external functions thus change the declarations to use C linkage when building the mobile runtime.

Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154/)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] update external functions for mobile build"


This is part of the changes to enable NNC AOT compilation for mobile.
The generated kernels need to call these external functions thus change the declarations to use C linkage when building the mobile runtime.

Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154/)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] update external functions for mobile build"


This is part of the changes to enable NNC AOT compilation for mobile.
The generated kernels need to call these external functions thus change the declarations to use C linkage when building the mobile runtime.

Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154/)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] update external functions for mobile build"


This is part of the changes to enable NNC AOT compilation for mobile.
The generated kernels need to call these external functions thus change the declarations to use C linkage when building the mobile runtime.

Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154/)

[ghstack-poisoned]

raziel commented on torch/csrc/jit/tensorexpr/external_functions.cpp:
```diff
@@ -196,6 +204,29 @@ void nnc_aten_mean(
+    int64_t* buf_dims,
+    int8_t* buf_dtypes,
+    int64_t args_num,
+    int64_t* extra_args) {
```
why not make this alfa and beta explicitly? is it just to avoid many declarations for now?

ljk53 replied:
> why not make this alfa and beta explicitly? is it just to avoid many declarations for now?

I think so - this way, the upstream caller side (IR Node -> external function call TensorExpr) can be consolidated into a simple shared helper function. @ZolotukhinM / @bertmaher might have more context.

It's a manual process to add new external functions today, and the overload resolution is ad-hoc, too (checkout how conv2d handles nullable bias param).

It's ok as we are only working with a small set of models for now. We should build a scalable way to do this - presumably via codegen from native_functions.yaml. The external_functions implementation can definitely be generated, maybe the IR node -> external_function part can be codegen-ed as well, then the signature here can be changed to align with aten op signature, which can reduce a little overhead of argument translation.

ZolotukhinM replied:
Correct, all the wrappers must have the same signature, that's why we use generic "extra_args" here. This way it works for addmm, conv2d, average_pool_2d and presumingly all other ops. I'd expect it to stay this way even if we switch to automatic generation of these functions, because having a single signature for all wrappers makes llvm codegen much simpler.

raziel commented on torch/csrc/jit/tensorexpr/external_functions.cpp:
```diff
@@ -196,6 +204,29 @@ void nnc_aten_mean(
+  const at::Tensor& z = tensors[3];
+  try {
+    at::addmm_out(r, x, y, z, extra_args[0], extra_args[1]);
+  } catch (...) {
```
The idea you were talking about in todays meeting would be to start with a simpler generic try/catch where these (aten based) external functions get called which would then raise the proper exception and debug handle.
Is that right?

ljk53 replied:
That's my original thought. @ZolotukhinM mentioned the exception handling is disabled in the generated assembly code, so I think the options are:
1) Figure out how to add exception handling capability in the generated llvm assembly - if I understand correctly just to pass through the exceptions thrown by these methods requires some change;
2) (suggested by @bertmaher ) make these functions to return error code / debug handle instead of throwing; We need codegen return-code check logic (and jump to the end of the function if error code is set) on the callsites of the assembly code. It's basically an ad-hoc exception handling.

kimishpatel replied:
So if external calls are not in try catch block do we just terminate?

kimishpatel replied:
So I did not quite understand the reason behind error handling mechanism in tensorexpr. But in TVM, which I guess is larger in scope, there is some dynamic error handling it seem. I did not understand fully but, presumably you can have exprs that have condition checks and they can throw if they fail.

When we have dynamic shape enabled I assume we will need to handle at least the shapes?

Some code from TVM that I am sure you guys know already:
https://codebrowser.bddppq.com/tvm/tvm/include/tvm/tir/stmt.h.html#tvm::tir::AssertStmtNode
https://codebrowser.bddppq.com/tvm/tvm/src/target/llvm/codegen_cpu.cc.html#911

cc: @ZolotukhinM @bertmaher

ZolotukhinM replied:
> So I did not quite understand the reason behind error handling mechanism in tensorexpr

Currently, there is no error handling mechanism in TE :) The reason why calls are guarded with catch-all blocks are not fundamental, that's a detail of the current implementation. Eventually we would like to add a support for error handling and as @ljk53 mentioned in the comment above we have two options for that: either 1) properly support exceptions in the generated code, or 2) go with the return code mechanism and assume that exceptions could be raised at the call-site of generated code (where we invoke NNC kernel). From implementation perspective, (2) seems much simpler and if the difference is not critical otherwise, I'd go that route.

ZolotukhinM replied:
> So if external calls are not in try catch block do we just terminate?

That's probably an undefined behavior or something: mark the wrappers (e.g. "nnc_aten_addmm") as `no-throw` in the generated LLVM code. I guess if that contract is violated random things could happen as LLVM would optimize the rest of the generated code under that assumption.

raziel commented on torch/csrc/jit/tensorexpr/external_functions.cpp:
```diff
@@ -196,6 +204,29 @@ void nnc_aten_mean(
+  }
+}
+
+#ifndef C10_MOBILE
```
Just curious, why are these registrations not needed for mobile?

ljk53 replied:
These registrations are used by the compiler at AOT compilation time. The registrations are not needed by the runtime as the assembly code will directly call the external functions without going through the registration table.

To include these we will need add a couple more source files to the runtime build, so I simply bypass them here.

C10_MOBILE implies mobile + runtime (no AOT / JIT compilation). Maybe we can create a dedicated macro for the runtime part (I'm little hesitate as new macros will add maintenance complexity).

ZolotukhinM replied:
These registrations are needed for llvm codegen to figure out what function to call to when it sees a "nnc_aten_conv2d" external call. I assume that codegen would be happening at the server side (since we're doing AOT) where all these registrations are present. Once the code is generated, we no longer need the registrations - thus in the mobile build we don't need them, we only need the functions themselves so that linker would find them.

ljk53 committed:
Update on "[pytorch][nnc] update external functions for mobile build"




This is part of the changes to enable NNC AOT compilation for mobile.
The generated kernels need to call these external functions thus change the declarations to use C linkage when building the mobile runtime.

Added nnc_aten_addmm external function.


Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154/)

Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] update external functions for mobile build"




This is part of the changes to enable NNC AOT compilation for mobile.
The generated kernels need to call these external functions thus change the declarations to use C linkage when building the mobile runtime.

Added nnc_aten_addmm external function.


Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154/)

Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/external_functions.cpp:
```diff
@@ -196,6 +204,29 @@ void nnc_aten_mean(
   }
 }
 
+void nnc_aten_addmm(
```
If we add a new external call function, let's add a test for it as well (they live in https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_external_calls.cpp).

Another option is to not include `addmm` in this PR at all - for this particular op it might make sense to decompose it to `add` + `mm`, both of which are already present.

ljk53 replied:
> If we add a new external call function, let's add a test for it as well (they live in https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_external_calls.cpp).

Test case added.

> Another option is to not include `addmm` in this PR at all - for this particular op it might make sense to decompose it to `add` + `mm`, both of which are already present.

@ZolotukhinM I feel this kinda of decomposition/translation would introduce complexity (and slightly more overhead). I prefer we either express an operator with pure TensorExpression or with straightforward extern function call. The operator schema in the TorchScript IR and in the ATen/Functions.h interface derive from the same source (native_functions.yaml). I think we should simply stick to the exact mapping and (eventually) automatically generate the boilerplate code.

ZolotukhinM replied:
I personally slightly prefer the other way (mainly because we reuse code rather than adding a ver similar piece), but I'm fine with either one. I also see your point about sticking to the 1-1 mapping, and it makes perfect sense to me as well. Btw, @Chillee was interested in looking into generating the ext-calls wrappers automatically too, so maybe we could start a discussion about it.

ljk53 replied:
P.S. just checked the ATen native cpu implementation, seems "matmul" calls "mm", which in turn calls "addmm", which then calls gemm implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L1094

So "addmm" this more "basic" in this regards, lol.

kimishpatel replied:
> if we can directly invoke QNNPack / XNNPack micro-kernels from NNC-generated assembly.

I am not sure how straight forward this is going to be. Mainly because there is a state to be created before. In TVM based approaches where equivalent would have been to actually write weight transformation op and then conv/gemm op. Then weight transformation op can be "precomputed" based on const propagation.

In this case I imagine we would have to have two microkernels. 1) weight prepacking 2) actual op. Then presume weight prepacking can be const propagated.

However, there are a few challenges with these approaches:
1. It introduces a tight coupling between NNC and QNNPACK/XNNPACK. QNNPACK fork being internal it might be relatively easy in practice but not in good in theory. The implementations of internals of QNNPACK op have now been exposed to outside and going forward we need to keep them in sync. For example if for particular kernel that produces and output tile of 8x8 for gemm but changes the packing of weights, will now require coordination with NNC. Doing this with XNNPACK will be even harder.
2.  For equivalent perf we need to have threading enabled the same way in XNNPACK/QNNPACK since NNC runtime will invoke ukernels differently compared to NNC. This is where efficiency of threadpool might matter too.

I think the easiest path for invoking these ops is based off of their prepacked variant/quantized variant. Like `prepacked::linear_pack`, `prepacked::linear_run` or `quantized::linear_prepack`, `quantized::linear_run`.

ZolotukhinM reviewed (approved):
Left one remark on `nnc_aten_addmm` function. Other than that, the PR looks good to me!

ljk53 committed:
Update on "[pytorch][nnc] update external functions for mobile build"




This is part of the changes to enable NNC AOT compilation for mobile.
The generated kernels need to call these external functions thus change the declarations to use C linkage when building the mobile runtime.

Added nnc_aten_addmm external function.


Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154/)

Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] update external functions for mobile build"




This is part of the changes to enable NNC AOT compilation for mobile.
The generated kernels need to call these external functions thus change the declarations to use C linkage when building the mobile runtime.

Added nnc_aten_addmm external function.


Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154/)

Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] update external functions for mobile build"




This is part of the changes to enable NNC AOT compilation for mobile.
The generated kernels need to call these external functions thus change the declarations to use C linkage when building the mobile runtime.

Added nnc_aten_addmm external function.


Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154/)

Differential Revision: [D27897154](https://our.internmc.facebook.com/intern/diff/D27897154)

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/tensorexpr/external_functions.cpp:
```diff
@@ -196,6 +204,29 @@ void nnc_aten_mean(
+  const at::Tensor& y = tensors[2];
+  const at::Tensor& z = tensors[3];
+  try {
+    at::addmm_out(r, x, y, z, extra_args[0], extra_args[1]);
```
Why addmm with 3 inputs?

kimishpatel commented on test/cpp/tensorexpr/test_external_calls.cpp:
```diff
@@ -227,6 +227,74 @@ TEST(ExternalCall, Conv2d_nobias_noargs) {
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+  std::vector<float> mat1_buf(100 * 200, 6.f);
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+  std::vector<float> mat2_buf(200 * 300, 11.f);
```
Why does this have to be done this way? Can we not extract the `mat1` tensors data pointer?

ZolotukhinM replied:
It works both ways. In fact, I think these arguments are converted to `void*` on line 284 and then a `CallArg` object is constructed from these pointers:
https://github.com/pytorch/pytorch/blob/42d073a7e9b338ee3d9224d525ca0383b6272af9/torch/csrc/jit/tensorexpr/codegen.h#L128

There is also a way to invoke codegen with `void*` pointers directly: #55113 (it was reverted but will be relanded soon).
